### PR TITLE
fix eos issue in finetune/generation

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
@sywangyi Can you give more information about this issue and how this PR fixes it please?

**Comment by sywangyi** [comment]
this is the jira issue description:
```
python ../gaudi_spawn.py --world_size 2 --use_mpi run_lora_clm.py \
--model_name_or_path google/gemma-2b-it --per_device_train_batch_size 16 --per_device_eval_batch_size 16 \
--do_train --do_eval --num_train_epochs 15 --output_dir ./output/2b_2hpu_16bs_15ep --save_total_limit 1 \
--gaudi_config_name Habana/gpt2 --use_habana --gradient_checkpointing --throughput_warmup_steps 3 --bf16 \
--logging_strategy epoch --evaluation_strategyepoch \
--lora_target_modules "q_proj" "o_proj" "k_proj" "v_proj" "gate_proj" "up_proj" "down_proj" --lora_rank=8 \
--lora_alpha=16 --lora_dropout=0.05 --dataset_name mamamiya405/finred
 ```
***** train metrics *****
epoch                       =        15.0
  max_memory_allocated (GB)   =92.57
  memory_allocated (GB)       =20.34
  total_flos                  = 856108315GF
  total_memory_available (GB) =       94.62
  train_loss                  =      0.3885
  train_runtime               =  0:43:57.18
  train_samples_per_second    =57.601
  train_steps_per_second      =1.803
 
***** Running Evaluation *****
100%|██████████| 94/94 [00:14<00:00,  5.96it/s]***** eval metrics *****
100%|██████████| 94/94 [00:22<00:00,  4.14it/s]
epoch                       =       15.0
  eval_accuracy               =     0.3903
  eval_loss                   =     0.4505
  eval_runtime                = 0:00:22.76
  eval_samples                =       3000
  eval_samples_per_second     =128.77
  eval_steps_per_second       =4.035
  max_memory_allocated (GB)   =92.57
  memory_allocated (GB)       =20.34
perplexity                  =      1.569
total_memory_available (GB) =      94.62
the output from LoRA trained model is

PROMPT

<bos><start_of_turn>user

Answer the below Query based on the Content given below. #### Query: Given phrases that describe the relationship between two words/phrases as options, extract the word/phrase pair and the corresponding lexical relationship between them from the input text. The output format should be 'relation1: word1, word2; relation2: word3, word4'. Options: product/material produced, manufacturer, distributed by, industry, position held, original broadcaster, owned by, founded by, distribution format, headquarters location, stock exchange, currency, parent organization, chief executive officer, director/manager, owner of, operator, member of, employer, chairperson, platform, subsidiary, legal form, publisher, developer, brand, business division, location of formation, creator. #### Content: PARIS, July 31 (Reuters) - European missile maker MBDA moved closer towards a reshuffle of ownership on Friday as Airbus Group signalledinterest in expanding its share, while Italy's Finmeccanica (Other OTC: FINMF - news ) said it would be open to selling 'at the right price'.<end_of_turn>

<start_of_turn>model

 OUTPUT

user

Answer the below Query based on the Content given below. #### Query: Given phrases that describe the relationship between two words/phrases as options, extract the word/phrase pair and the corresponding lexical relationship between them from the input text. The output format should be 'relation1: word1, word2; relation2: word3, word4'. Options: product/material produced, manufacturer, distributed by, industry, position held, original broadcaster, owned by, founded by, distribution format, headquarters location, stock exchange, currency, parent organization, chief executive officer, director/manager, owner of, operator, member of, employer, chairperson, platform, subsidiary, legal form, publisher, developer, brand, business division, location of formation, creator. #### Content: PARIS, July 31 (Reuters) - European missile maker MBDA moved closer towards a reshuffle of ownership on Friday as Airbus Group signalled interest in expanding its share, while Italy's Finmeccanica (Other OTC: FINMF - news ) said it would be open to selling 'at the right price'.

model

According to the query, the following is the relationship between Airbus Group and MBDA in the context of the article;

 

operator: MBDA, Airbus Group

 

Therefore, the word pair is: operator: MBDA, Airbus Group; product_or_material_produced: Airbus, aircraft; product_or_material_produced: MBDA, aircraft systems; industry: Airbus, aircraft manufacturer; industry: MBDA, aircraft manufacturer;owned_by: , Airbus Group; owned_by: , MBDA; headquarters_location: Airbus, Paris; headquarters_location: MBDA, Paris; founded_by: , Airbus Group; founded_by: , MBDA; chief_executive_officer: Airbus, Jean-Michel Martin

The expected output is

user

Answer the below Query based on the Content given below. #### Query: Given phrases that describe the relationship between two words/phrases as options, extract the word/phrase pair and the corresponding lexical relationship between them from the input text. The output format should be 'relation1: word1, word2; relation2: word3, word4'. Options: product/material produced, manufacturer, distributed by, industry, position held, original broadcaster, owned by, founded by, distribution format, headquarters location, stock exchange, currency, parent organization, chief executive officer, director/manager, owner of, operator, member of, employer, chairperson, platform, subsidiary, legal form, publisher, developer, brand, business division, location of formation, creator. #### Content: PARIS, July 31 (Reuters) - European missile maker MBDA moved closer towards a reshuffle of ownership on Friday as Airbus Group signalled interest in expanding its share, while Italy's Finmeccanica (Other OTC: FINMF - news ) said it would be open to selling 'at the right price'.

model

owned_by: MBDA, Airbus Group; owner_of: Airbus Group, MBDA

 

Test script is


```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
from peft import PeftModel
from optimum.habana.transformers.modeling_utils import adapt_transformers_to_gaudi
from optimum.habana.utils import set_seed

adapt_transformers_to_gaudi()
model_id = "google/gemma-2b-it"
peft_id = "output/2b_2hpu_16bs_15ep"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16)
model = PeftModel.from_pretrained(model, peft_id, torch_dtype=model.dtype)
model = model.eval().to("hpu")
set_seed(0)

chat = [
    { "role": "user", "content": "Answer the below Query based on the Content given below. #### Query: Given phrases that describe the relationship between two words/phrases as options, extract the word/phrase pair and the corresponding lexical relationship between them from the input text. The output format should be 'relation1: word1, word2; relation2: word3, word4'. Options: product/material produced, manufacturer, distributed by, industry, position held, original broadcaster, owned by, founded by, distribution format, headquarters location, stock exchange, currency, parent organization, chief executive officer, director/manager, owner of, operator, member of, employer, chairperson, platform, subsidiary, legal form, publisher, developer, brand, business division, location of formation, creator. #### Content: PARIS, July 31 (Reuters) - European missile maker MBDA moved closer towards a reshuffle of ownership on Friday as Airbus Group signalled interest in expanding its share, while Italy's Finmeccanica (Other OTC: FINMF - news ) said it would be open to selling 'at the right price'." },
]

prompt = tokenizer.apply_chat_template(chat, tokenize=False, add_generation_prompt=True)
print(prompt)
inputs = tokenizer.encode(prompt, add_special_tokens=False, return_tensors="pt")
outputs = model.generate(input_ids=inputs.to(model.device), max_new_tokens=150, lazy_mode=False, hpu_graphs=False, ignore_eos=False)
print(tokenizer.decode(outputs[0], skip_special_tokens=True))

```

**Comment by sywangyi** [comment]
fix issues in this PR.
1. The eos is not learned correctly.  The default add_eos_token of tokenizer is false. if dataset_concatenation is not set, the currently logic will not add eos at the end of sequence.
2.  The inference use chat template. but run_lora_clm.py has no such template, it's using alpaca instruction format instead. 
3.  The eval accuracy calculation does not mask padding label. make the eval accuracy low.
4.  chat template has bos, should not add duplicated bos in tokenizer if it's chat.

**Comment by sywangyi** [comment]
After the fix,  the output meet the expectation.
owned_by: MBDA, Airbus Group; owner_of: Airbus Group, MBDA

File: examples/language-modeling/run_lora_clm.py [hunk 1]
```diff
@@ -263,6 +263,10 @@ class DataArguments:
         default=False,
         metadata={"help": "Whether to have a SQL style prompt"},
     )
+    chat_prompt: bool = field(
+        default=False,
+        metadata={"help": "Whether to have a chat style prompt"},
```
Comment (yafshar) [inline examples/language-modeling/run_lora_clm.py:268]:
```suggestion
        metadata={"help": "Whether to have a chat style prompt."},
```

Please add the period for sql_prompt metadata as well.

File: examples/language-modeling/run_lora_clm.py [hunk 2]
```diff
@@ -676,19 +700,25 @@ def main():
     if tokenizer.pad_token_id is None:
         tokenizer.pad_token_id = tokenizer.eos_token_id
 
-    def tokenize(prompt, add_eos_token=True):
+    def tokenize(prompt, add_eos_token=True, add_bos_token=True):
+        add_eos_token_o = tokenizer.add_eos_token
+        add_bos_token_o = tokenizer.add_bos_token
         if not data_args.dataset_concatenation:
-            add_eos_token = False
+            tokenizer.add_eos_token = add_eos_token
```
Comment (yafshar) [inline examples/language-modeling/run_lora_clm.py:707]: @sywangyi what about the else condition? Does `tokenizer.add_eos_token` exist?

**Comment by yafshar** [comment]
@sywangyi, can you try the test case with `ignore_eos=True)` ? Is the behavior as expected by you

File: examples/language-modeling/run_lora_clm.py (hunk 2 above)
Comment (sywangyi) [reply examples/language-modeling/run_lora_clm.py:707]: the model tokenizer class contain such information. like https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/tokenization_llama_fast.py#L135-L136  and I do not want to override this value. since when dataset_concatenation is on, we will add eos manually, we only need tokenizer to add eos when dataset_concatenation is off.

**Comment by sywangyi** [comment]
> @sywangyi, can you try the test case with `ignore_eos=True)` ? Is the behavior as expected by you

no, the issue is the model does not learn the eos info during finetune, so output could not be ended as expected even ignore_eos is True during text generation.

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss can you pleases check this PR.

**Review by regisss — APPROVED** [review]
